Exporting ioctl

uraimo opened the pull request:
#### What's in this pull request?

This pull request exports the variadic `ioctl` system call, needed to interact with device files on Linux and other Unix-like systems. At the moment, Linux Swift libraries that interact with a device driver need to implement a c wrapper for `ioctl` in a separate clang/llvm module (essentially the same c module is duplicated in every library that does something with devices). 

Having `ioctl` available in Swift would allow the creation of pure-Swift device libraries (more portable, easier to compile).

The two new functions contained in this PR follow the same approach of the already exported `fcntl`, one function with an int parameter and one with a pointer. But unlike `fcntl` i don't expect `ioctl` to be used internally in Foundation any time soon.

On the testing side, the only thing that can be tested across platforms is the presence of the function, since aside from a few obscure commands related to tty devices, ioctl requests are heavily platform-dependent. I'm including a single test that tries an ioctl on a non-existing file (failure is expected).   

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

modocache commented:
@swift-ci please test

jrose-apple commented on stdlib/public/Platform/Misc.c:
```diff
@@ -49,6 +50,16 @@ extern int _swift_Platform_fcntlPtr(int fd, int cmd, void* ptr) {
 
+extern int
+_swift_Platform_ioctl(int fd, unsigned long int request, int value) {
+  return fcntl(fd, request, value);
```
Is this supposed to be ioctl?

uraimo replied:
Yes, fixed and testing it locally.

uraimo replied:
Tests passed on OSX&Linux.

jrose-apple commented:
Is there anything that can be tested across *most* Linux systems? I'd be fine with a `#if` around that part of the test if it meant we got a little more guarantee that things were actually working.

uraimo commented:
I've updated the PR:
* Fixed the fcntl/ioctl typo
* Added missing SWIFT_CC(swift)
* Removed @warn_unused_result
* Added a Linux-only test that retrieves the interface addresses for a newly created socket (should work on every Linux and probably on BSDs too but i put it inside an #if(Linux) anyway)

Tested locally on both OSX and Linux (Ubuntu 15).

jrose-apple commented:
Seems okay to me. @modocache, @gribozavr?

@swift-ci Please test

gribozavr commented:
@uraimo Thank you!  What about an overload for commands without an argument?

gribozavr commented on test/1_stdlib/POSIX.swift:
```diff
@@ -82,7 +82,40 @@ POSIXTests.test("sem_open existing O_EXCL fail") {
+    var vptr = UnsafeMutablePointer<Void>(ptr)  
+    let io = ioctl(sock, UInt(SIOCGIFCONF), vptr);
+    expectGE(io, 0)
+  })
```
I think you should be able to do this:

```swift
  var ic = ifconf()
  let io = ioctl(sock, UInt(SIOCGIFCONF), &ic);
  expectGE(io, 0)
```

uraimo replied:
Yep, way better, that was a remnant of previous ioctl attempts.

uraimo commented:
Makes sense, added.

gribozavr commented:
@swift-ci Please test and merge
